### PR TITLE
Custom equality and comparison for F# lists

### PR DESCRIPTION
**Pull request by cartermp**
This is mostly just for my curiosity at this point

#1838 is an annoying problem because F# lists are DUs, and because of that the generated hash function for them builds up a giant tree-like structure. This is probably the correct thing to do for most DUs, but lists are special in that people don't really use them "as discriminated unions", but really just collections of things. This results in problems like trying to hash a large list (e.g., 50k items) and getting a stack overflow because the generated hash function is absolutely enormous. Solving this general case for all DUs to be performant with very large structures (i.e., not just lists) looks like it would be very difficult. Probably better to just special case lists and move on, since they're the most affected by the issue anyways.

This PR changes the list type to define its own structural equality and comparison. Some basic testing with FSI was done, and some of the unit tests that can run within VS were ran successfully. But we'll see what happens with _P E R L T E S T I N G_ and other stuff that doesn't show up in standard IDE tools.

The ramifications of this change that I know of:

* Any code that depends on a specific value generated from a hash code (i.e., incorrect code) will break
* Any code that depends on the list type having the `StructuralEquality` and `StructuralComparison` attributes via reflection will rbeak

### CODE REVIEW

**Comment by Smaug123** [comment]
I don't know if the following suggestion is too much complexity for too little gain, but: how about some kind of "CheaplyHashable" inference, where some fixed known types like int and string are CheaplyHashable. Then we could do a bit better than this: we could (in theory) tail-recursively compute a hash for any data structure with only one non-CheaplyHashable element without having to go into continuation-passing style.

**Comment by abelbraaksma** [comment]
@Smaug123, you mean where the type to be hashed is itself a recursive type, but any of its dependencies are non recursive?

File: src/fsharp/FSharp.Core/prim-types.fsi [hunk 1]
```diff
@@ -1948,6 +1948,7 @@ namespace Microsoft.FSharp.Collections
         interface IEnumerable
         interface IReadOnlyCollection<'T>
         interface IReadOnlyList<'T>
+        interface IComparable
```
Comment (0x53A) [inline src/fsharp/FSharp.Core/prim-types.fsi:1951]: At the moment List implements a whole bunch of interfaces, not just IComparable:

image

**Comment by Smaug123** [comment]
> @Smaug123, you mean where the type to be hashed is itself a recursive type, but any of its dependencies are non recursive?

Yep: basically if we can infer that the type is isomorphic to "a list, plus any number of small extra leaves at each level".

**Comment by cartermp** [comment]
Just noting that this is a draft still; for some reason CI would not kick off unless I maked it as not a draft.

The surface area failures are due to this:

image
image

Additionally, the equals/hashcode/etc. methods are also marked as sealed overrides, which you can't write in F# code.

I'm a little lost as to what to do here, since I'm not sure there's a way to define the List type with these things explicitly and have it match 1:1 with what the compiler generates today

**Comment by cartermp** [comment]
I'm going to close this out. If anyone else is interested in pursuing it, feel free to ping me if you have any questions.

I don't know how to consider the changes that this PR would imply. There is a difference in what the compiler generates with what I can express explicitly on the type and it gives me _the spookums_ to think about how that could affect existing F# code.

**Pull request closed without merging**